fastchat/serve/vllm_worker.py

```diff
@@ -90,27 +94,54 @@ async def generate_stream(self, params):
         top_p = max(top_p, 1e-5)
         if temperature <= 1e-5:
             top_p = 1.0
-        sampling_params = SamplingParams(
-            n=1,
-            temperature=temperature,
-            top_p=top_p,
-            use_beam_search=False,
-            stop=list(stop),
-            max_tokens=max_new_tokens,
-        )
-        results_generator = engine.generate(context, sampling_params, request_id)
-
-        async for request_output in results_generator:
-            prompt = request_output.prompt
-            if echo:
-                text_outputs = [
-                    prompt + output.text for output in request_output.outputs
-                ]
-            else:
-                text_outputs = [output.text for output in request_output.outputs]
-            text_outputs = " ".join(text_outputs)
-            # Note: usage is not supported yet
-            ret = {"text": text_outputs, "error_code": 0, "usage": {}}
+        try:
+            sampling_params = SamplingParams(
+                n=n,
+                temperature=temperature,
+                top_p=top_p,
+                use_beam_search=use_beam_search,
+                stop=list(stop),
+                max_tokens=max_new_tokens,
+                best_of=best_of,
+            )
+
+            results_generator = engine.generate(context, sampling_params, request_id)
+
+            async for request_output in results_generator:
+                prompt = request_output.prompt
+                prompt_tokens = len(request_output.prompt_token_ids)
+                for out in request_output.outputs:
+                    completion_tokens = len(out.token_ids)
+
+                output_usage = {
+                    "prompt_tokens": prompt_tokens,
+                    "completion_tokens": completion_tokens,
+                    "total_tokens": prompt_tokens + completion_tokens,
+                }
+
+                text_outputs = []
+                finish_reason = []
+                for output in request_output.outputs:
+                    if echo:
+                        text_outputs.append(prompt + output.text)
+                    else:
+                        text_outputs.append(output.text)
+
+                    finish_reason.append(output.finish_reason)
+
+                ret = {
+                    "text": text_outputs,
```

Comment:
It seems you still make the output format of `vllm_worker.py` and `model_worker.py` different.
Although you correctly handled this in `openai_api_server.py`, there are other places you need to update.

There are some options

1. Update both model_worker.py and vllm_worker.py to make their output format the same
2. Make all places that talk to a model worker able to handle different output formats (e.g., openai_api_server.py, gradio_web_server.py)
3. Only use this list format when beam search is enabled.

Reply:
New patch now keep original n=1 logic, but only add beam_search and best_of option.
It seems to me that n>1 is truely rarely used, so temp remove it from original patch.